044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,174 @@
+This combined with the varying readiness checks means we need to change the way the KafkaRoller works:
+
+* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
+* If more than one controller pod has not become ready, KafkaRoller should try to roll the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to RUNNING, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
+
+## Proposed order of tasks to complete this proposal
+
+Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:
+
+### Phase 1 - Strimzi only supports combined mode
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090. 
+This is an improvement on the current state where the pod is marked as alive and ready as soon as it starts up.
+* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
+This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
+
+### Phase 2 - Strimzi adds support for broker only and controller only modes
+* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
+* Broker only pods are marked as both alive and ready if listening on port 9091.
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).
```

Comment:
These will likely be shell scripts -> so do they need to wait for the actual support for the broker and controller modes? OR can they be added right away just not used for the time being?

Reply:
The scripts could be written in advance, however these phases are meant to describe the actual checks that can be used, so if we did write them in advance they would have to be disabled.

I think the advantage of waiting is we will need a mechanism to determine whether the node is running in broker only or controller only or combined mode, and it might be that once we have added support for those modes it becomes clearer what mechanism that is. For example you could check the server properties, but it might be we end up needing an env var on the pod, and if so it would be simpler to use that.